oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/CompositeNodeStoreLuceneIndexTest.java

```diff
@@ -0,0 +1,506 @@
+                containsString("/* lucene:luceneTest3(/oak:index/luceneTest3) foo3:bar"));
+
+        result = compositeRepoV2.getCompositeQueryManager().createQuery("/jcr:root//*[@foo3 = 'bar'] order by @jcr:path", "xpath").execute();
+
+        // Check that proper nodes are returned by the query even after traversal from both readonly version 2 and global read write parts
+        assertEquals("/content-foo3/node-0, /content-foo3/node-1, " +
+                "/libs/node-luceneTest3-0, /libs/node-luceneTest3-1, /libs/node-luceneTest3-2", getResult(result,"jcr:path"));
+
+
+
+        // Now check that the V1 instance still uses B for foo2 , A for foo and traverses for foo3
+
+        result = compositeRepoV1.getCompositeQueryManager().createQuery("explain /jcr:root//*[@foo = 'bar']", "xpath").execute();
+        assertThat(result.getRows().next().toString(),
+                containsString("/* lucene:luceneTest(/oak:index/luceneTest) foo:bar"));
+
+        result = compositeRepoV1.getCompositeQueryManager().createQuery("/jcr:root//*[@foo = 'bar'] order by @jcr:path", "xpath").execute();
+
+        assertEquals("/content-foo/node-0, /content-foo/node-1, " +
+                "/libs/node-luceneTest-0, /libs/node-luceneTest-1, /libs/node-luceneTest-2", getResult(result,"jcr:path"));
+
+        // foo 2 check
+
+        result = compositeRepoV1.getCompositeQueryManager().createQuery("explain /jcr:root//*[@foo2 = 'bar']", "xpath").execute();
+        assertThat(result.getRows().next().toString(),
+                containsString("/* lucene:luceneTest2(/oak:index/luceneTest2) foo2:bar"));
+
+        result = compositeRepoV1.getCompositeQueryManager().createQuery("/jcr:root//*[@foo2 = 'bar'] order by @jcr:path", "xpath").execute();
+
+        assertEquals("/content-foo2/node-0, /content-foo2/node-1, " +
+                "/libs/node-luceneTest2-0, /libs/node-luceneTest2-1, /libs/node-luceneTest2-2", getResult(result,"jcr:path"));
+
+        // foo 3 check
+        compositeRepoV1.login();
+        result = compositeRepoV1.getCompositeQueryManager().createQuery("explain /jcr:root//*[@foo3 = 'bar']", "xpath").execute();
+        assertThat(result.getRows().next().toString(),
+                containsString("/* traverse \"//*\" where ([a].[foo3] = 'bar'"));
+
+        result = compositeRepoV1.getCompositeQueryManager().createQuery("/jcr:root//*[@foo3 = 'bar'] order by @jcr:path", "xpath").execute();
+
+        assertEquals("/content-foo3/node-0, /content-foo3/node-1", getResult(result,"jcr:path"));
+
+        compositeRepoV1.cleanup();
+        compositeRepoV2.cleanup();
+    }
+
+    @After
+    public void cleanup() throws Exception {
+        for (JackrabbitSession session : sessions) session.logout();
+        for (Repository repo : repositories) shutdown(repo);
+    }
+
+    private void setupIndexAndContentInComposite(String indexName, String indexedProperty, CompositeRepo compositeRepo) throws Exception {
```

Comment:
Could this be a method of the CompositeRepo class?

Reply:
This was moved to compositeRepo